### PR TITLE
introduction of wavefront.proxy.lograwdata

### PR DESCRIPTION
**Pull request by howardyoo**
which can be used to enable proxy to log ACTUAL data received by channel string handler, before any pre-processing or formatting occurs. This feature is useful when you are trying to troubleshoot or test pre-processor rules or formatting of data (e.g. Graphite) to compare the original data against the finished point data ready to be submitted to Wavefront server.

### CODE REVIEW

File: proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java [hunk 1]
```diff
@@ -69,6 +79,9 @@ protected void channelRead0(ChannelHandlerContext ctx, String msg) throws Except
       }
     }
     // if msg does not match metadata keywords then treat it as a point
+    if(logRawDataFlag == true) {
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java]: if (logRawDataFlag)

File: proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java [hunk 2]
```diff
@@ -46,6 +48,10 @@ public ChannelStringHandler(Decoder<String> decoder,
     this.pointHandler = pointhandler;
     this.preprocessor = preprocessor;
     this.metadataHandler = metadataHandler;
+
+    // check the property setting for logging raw data
+    String logRawDataProperty = System.getProperty("wavefront.proxy.lograwdata");
+    logRawDataFlag = logRawDataProperty != null && logRawDataProperty.equalsIgnoreCase("true");
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java]: See vasily's commit, use a double instead from 0 to 1 to and a rate limiter to have control over how much to log

Comment (howardyoo) [reply proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java]: got it, working

**Comment by howardyoo** [comment]
new code committed, checks passed.

**Comment by basilisk487** [comment]
@howardyoo suggestion: since this is done in ChannelStringHandler that has remote context, prepend the point being logged with a remote IP address in square braces, this can be really useful in identifying where certain points are coming from

**Comment by howardyoo** [comment]
@basilisk487 , Sounds like a very good idea. Will apply it to the commit. Thanks!

**Comment by howardyoo** [comment]
Applied change to include ip address and port information of the remote host sending the metric.

File: proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java [hunk 3]
```diff
@@ -46,15 +53,27 @@ public ChannelStringHandler(Decoder<String> decoder,
...
+
+    // make sure the rate fits between 0.0d - 1.0d
+    if(logRawDataRate < 0.0d) {
+      logRawDataRate = 0.0d;
+    }
+    else if(logRawDataRate > 1.0d) {
+      logRawDataRate = 1.0d;
+    }
+    if(logRawDataRate > 0.0d) {
+      rawDataLogger.info("Raw data logging is enabled with " + (logRawDataRate * 100) + "% sampling");
```
Comment (basilisk487) [inline proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java]: Suggestion: making this configurable also through rawDataLogger.isLoggable(Level.FINEST) would make it possible to enable and disable the logger on the fly without restarting the proxy

File: proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java [hunk 4]
```diff
@@ -69,6 +88,17 @@ protected void channelRead0(ChannelHandlerContext ctx, String msg) throws Except
       }
     }
     // if msg does not match metadata keywords then treat it as a point
+    // use data rate to determine sampling rate
+    // logging includes the source host and port
+    if(logRawDataRate >= 1.0d || (logRawDataRate > 0.0d && RANDOM.nextDouble() < logRawDataRate)) {
+      if(ctx.channel().remoteAddress() != null) {
+        String host = ((InetSocketAddress) ctx.channel().remoteAddress()).getHostName();
+        int port = ((InetSocketAddress) ctx.channel().remoteAddress()).getPort();
```
Comment (basilisk487) [inline proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java]: I believe logging remote address and local port would make more sense - remote port is not very useful once the session is over. Maybe write both [local port : remote addr : remote port]?

File: proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java [hunk 5]
```diff
@@ -69,6 +88,17 @@ protected void channelRead0(ChannelHandlerContext ctx, String msg) throws Except
       }
     }
     // if msg does not match metadata keywords then treat it as a point
+    // use data rate to determine sampling rate
+    // logging includes the source host and port
+    if(logRawDataRate >= 1.0d || (logRawDataRate > 0.0d && RANDOM.nextDouble() < logRawDataRate)) {
+      if(ctx.channel().remoteAddress() != null) {
+        String host = ((InetSocketAddress) ctx.channel().remoteAddress()).getHostName();
```
Comment (basilisk487) [inline proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java]: getHostName() is a reverse DNS lookup, which is a very expensive operation. In high traffic environments one getHostName() _per incoming connection_ has caused issues previously, so one rDNS lookup per logged point is virtually guaranteed to cause performance issues (and unleash the wrath of your network engineer as an added bonus).
Just logging remote IP (ctx.channel().remoteAddress().getAddress().getHostAddress()) should be fine, or cache rDNS lookup results like in PushAgent.reverseDnsCache

File: proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java [hunk 6]
```diff
@@ -86,6 +116,9 @@ public static void processPointLine(final String message,
     String pointLine = message.trim();
     if (pointLine.isEmpty()) return;
 
+
+
+
```
Comment (basilisk487) [inline proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java]: nit-pick: too many empty lines

File: proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java (hunk 5 above)
Comment (howardyoo) [reply proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java]: Ok, thanks for the comment. makes sense.

File: proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java (hunk 6 above)
Comment (howardyoo) [reply proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java]: Neat guy!

File: proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java (hunk 5 above)
Comment (howardyoo) [reply proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java]: changed the delimiter ':' to '>', as the local port could be confused as remote port. Removed remote port in the log.

**Comment by panghy** [comment]
@howardyoo do you have time to address vasily's comments?

**Comment by howardyoo** [comment]
Hi Clement! I've already made the changes, and submitted a PR on it!
Howard

**Comment by basilisk487** [comment]
@howardyoo the only thing missing is real-time reconfiguration, so you need to check for rawDataLogger.isLoggable(Level.FINEST) in the main method body, not in the constructor (and, preferably, not for every point - once a second is good)

**Comment by howardyoo** [comment]
Oh, okay. Will work on it :-).
Thanks,
Howard

**Comment by panghy** [comment]
Please resolve conflicts and for some reason I am not seeing a travis pass

File: proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java (hunk 3 above)
Comment (howardyoo) [reply proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java]: Applied to check the isLoggable - make it possible to dynamically enable/disable the feature.

File: proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java [hunk 7]
```diff
@@ -92,11 +93,12 @@ protected void channelRead0(ChannelHandlerContext ctx, String msg) throws Except
...
       if(ctx.channel().remoteAddress() != null) {
-        String host = ((InetSocketAddress) ctx.channel().remoteAddress()).getHostName();
-        int port = ((InetSocketAddress) ctx.channel().remoteAddress()).getPort();
-        rawDataLogger.info(String.format("[%s:%d]%s", host, port, msg));
+        String hostAddress = ((InetSocketAddress) ctx.channel().remoteAddress()).getAddress().getHostAddress();
+        int localPort = ((InetSocketAddress) ctx.channel().localAddress()).getPort();
+        rawDataLogger.info(String.format("[%s>%d]%s", hostAddress, localPort, msg));
       } else {
-        rawDataLogger.info(String.format("[:]%s", msg));
+        int localPort = ((InetSocketAddress) ctx.channel().localAddress()).getPort();
```
Comment (howardyoo) [inline proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java]: also changed from getting remote to local

File: proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java [hunk 8]
```diff
@@ -92,11 +93,12 @@ protected void channelRead0(ChannelHandlerContext ctx, String msg) throws Except
...
+        rawDataLogger.info(String.format("[>%d]%s", localPort, msg));
```
Comment (howardyoo) [inline proxy/src/main/java/com/wavefront/agent/ChannelStringHandler.java]: also, logging the ip and port - included in brackets.

**Comment by howardyoo** [comment]
@panghy , I've committed all the changes per vasily's request. See https://github.com/wavefrontHQ/java/pull/250/commits/34356b90b1cdb517ca3cd3bd9bf7236db090b21b
and 
https://github.com/wavefrontHQ/java/pull/250/commits/3bec4fc304e69d300cee4d6f7c18cb98ab1bc71a

there is no additional work left for me, so am waiting for all these would be committed.

**Comment by panghy** [comment]
Please rebase and also do a reformat

**Comment by howardyoo** [comment]
Ok, thanks!

**Comment by howardyoo** [comment]
Hi @panghy , is there any documents regarding code formatting that I can take a look? Also, what wold be the recommended way to safely rebase and re-commit the changes to the recent master branch?

**Comment by panghy** [comment]
Our code style settings is in https://github.com/sunnylabs/system/blob/master/.idea/codeStyleSettings.xml

This can be checked into every public repo that we have.

**Comment by howardyoo** [comment]
Thanks!

**Comment by howardyoo** [comment]
Closing this PR due to unnecessary number of commits during rebasing the branch. Don't want to commit 90 items just to make this change in a single java class. Will be creating a new PR based on the most recent master stream - fork it, and then branch again.

**Pull request closed without merging**